fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SizeStatisticsCollectorCursor.java

```diff
@@ -0,0 +1,642 @@
+     * @param continuation the starting cursor location to start stats collection
+     *
+     * @return a statistics collector of the given subspace
+     */
+    @Nonnull
+    public static SizeStatisticsCollectorCursor ofSubspace(@Nonnull Subspace subspace, @Nonnull FDBRecordContext context,
+                                                           @Nonnull ScanProperties scanProperties, @Nullable byte[] continuation) {
+        return new SizeStatisticsCollectorCursor(new SubspaceProviderBySubspace(subspace), context, scanProperties, continuation);
+    }
+
+    /**
+     * Encapsulates the distribution statistics returned by a SizeStatisticsCollectorCursor.
+     */
+    public static class SizeStatisticsResults {
+        private long keyCount;
+        private long keySize;
+        private long maxKeySize;
+        private long valueSize;
+        private long maxValueSize;
+        private long[] sizeBuckets;
+
+        private SizeStatisticsResults() {
+            this.keyCount = 0;
+            this.keySize = 0;
+            this.maxKeySize = 0;
+            this.valueSize = 0;
+            this.maxValueSize = 0;
+            this.sizeBuckets = new long[Integer.SIZE];
+        }
+
+        /**
+         * Update partial kv stats.
+         *
+         * @param kv latest key and value for partial stats update
+         */
+        private void updateStatistics(@Nonnull KeyValue kv) {
+            this.keyCount += 1;
+            this.keySize += kv.getKey().length;
+            this.maxKeySize = Math.max(this.maxKeySize, kv.getKey().length);
+            this.valueSize += kv.getValue().length;
+            this.maxValueSize = Math.max(this.maxValueSize, kv.getValue().length);
+            int totalSize = kv.getKey().length + kv.getValue().length;
+            if (totalSize > 0) {
+                this.sizeBuckets[Integer.SIZE - Integer.numberOfLeadingZeros(totalSize) - 1] += 1;
+            }
+        }
+
+        /**
+         * Sets partial kv stats from a protobuf serialization of partial stats results.
+         *
+         * @param partialResults partial kv stats from a protobuf serialization of partial stats results.
+         */
+        private void setFromProto(RecordCursorProto.SizeStatisticsPartialResults partialResults) {
+            this.keyCount = partialResults.getKeyCount();
+            this.keySize = partialResults.getKeySize();
+            this.maxKeySize = partialResults.getMaxKeySize();
+            this.valueSize = partialResults.getValueSize();
+            this.maxValueSize = partialResults.getMaxValueSize();
+            List<Long> bucketList = partialResults.getSizeBucketsList();
+            Arrays.setAll(sizeBuckets, i -> bucketList.get(i));
+        }
+
+        /**
+         * Gets protobuf representation of partial stats.
+         *
+         * @return protobuf representation of partial stats.
+         */
+        private RecordCursorProto.SizeStatisticsPartialResults getProto() {
```

Comment:
Could this be `toProto` to be consistent with our other classes?

Reply:
No problem